src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/Serializer/ResourceProtoSerializer.cs

```diff
@@ -0,0 +1,58 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+using OpenTelemetry.Resources;
+
+namespace OpenTelemetry.Exporter.OpenTelemetryProtocol.Implementation.Serializer;
+
+internal static class ResourceProtoSerializer
+{
+    private const int ReserveSizeForLength = 4;
+
+    internal static int WriteResource(byte[] buffer, int writePosition, Resource? resource)
+    {
+        OtlpProtoTagWriter.OtlpTagWriterState otlpTagWriterState = new OtlpProtoTagWriter.OtlpTagWriterState
+        {
+            Buffer = buffer,
+            WritePosition = writePosition,
+        };
+
+        if (resource != null && resource != Resource.Empty)
+        {
+            otlpTagWriterState.WritePosition = ProtobufSerializer.WriteTag(otlpTagWriterState.Buffer, otlpTagWriterState.WritePosition, FieldNumberConstants.ResourceSpans_Resource, ProtobufWireType.LEN);
+            int resourceLengthPosition = otlpTagWriterState.WritePosition;
+            otlpTagWriterState.WritePosition += ReserveSizeForLength;
+
+            foreach (KeyValuePair<string, object> attribute in resource.Attributes)
```

Comment:
This is going to allocate an enumerator because Resource.Attributes is `IEnumerable<KeyValuePair<string, object>>`. Internally it is a `List<...>`. So if you want to avoid the allocation you could do the trick...

```csharp
if (resource.Attributes is IReadOnlyList<KeyValuePair<string, object>> list)
{
   for (int i = 0; i < list.Count, i++)
   {
      var attribute = list[i];
      ...
   }
}
else 
{
      foreach (var attribute in resource.Attributes)
      {
         ...
      }
}
```